Improvements for the FSLeyes plugin

Drulex opened the pull request:
Detailed list of improvements below:
- Added "stop" functionality to `ProgressDialog` and hooked to `SCTCallThread` to send a `SIGINT` if the user presses stop.
- Improve `ErrorDialog` to display a specific/more useful error message to the user instead of the generic one (dialog shown when `SCTCallThread` returns non-zero codes).
- Refactor `SCTCallThread` to retrieve progress in real time.
- Added a `TxtCtrl` widget to the UI in order to display the `SCTCallThread` progress in real time.
- Various other small fixes.

Relates to:
- #2594

Screenshots:

**When processing:**
![image](https://user-images.githubusercontent.com/7751484/84798693-40487880-afc9-11ea-8c23-613ea8c5d70f.png)

**When command was successful:**
![image](https://user-images.githubusercontent.com/7751484/84798742-4dfdfe00-afc9-11ea-9546-ac318f6d6d10.png)

**When an error occurred:**
![image](https://user-images.githubusercontent.com/7751484/84798763-548c7580-afc9-11ea-98ea-08bdb3c1c423.png)

Drulex commented:
There are still a few minor TODOs that remain, but I wanted to implement the most useful features/fixes in the shortest time to improve the UX. I think that this PR makes sense as a first pass of improvements at least.

Drulex commented:
@zougloub can you do a quick sanity check? I used unix pipes for non-blocking reading of the streams (Windows support is provided via WSL https://github.com/neuropoly/spinalcordtoolbox/pull/2643)

jcohenadad commented:
@Drulex currently testing your PR. Overall, the new features look exciting!!! esp. the terminal output embedded in the FSLeyes, i love it!

i'm running into a minor issue and i was just wondering if this is expected behaviour: when pressing stop during a process, a window with no message appear, see here:

[Untitled.mov.zip](https://github.com/neuropoly/spinalcordtoolbox/files/4794694/Untitled.mov.zip)

jcohenadad commented:
another comment: when trying to run a function with not enough inputs, user should be informed (e.g. via a popup), e.g. see here: 
[sct-fsleyes_not_enough_inputs.mov.zip](https://github.com/neuropoly/spinalcordtoolbox/files/4794709/sct-fsleyes_not_enough_inputs.mov.zip)

Drulex commented:
> @Drulex currently testing your PR. Overall, the new features look exciting!!! esp. the terminal output embedded in the FSLeyes, i love it!
> 
> i'm running into a minor issue and i was just wondering if this is expected behaviour: when pressing stop during a process, a window with no message appear, see here:
> 
> [Untitled.mov.zip](https://github.com/neuropoly/spinalcordtoolbox/files/4794694/Untitled.mov.zip)

The window shows the `stderr` output if an error occurred. When I try this I get a traceback for a `KeyboardInterrupt` error (see attached)
[KeyboardInterrupt.mp4.zip](https://github.com/neuropoly/spinalcordtoolbox/files/4794729/KeyboardInterrupt.mp4.zip)

Maybe its OS related or maybe in that specific sequence you pressed stop too fast?

Drulex commented:
> another comment: when trying to run a function with not enough inputs, user should be informed (e.g. via a popup), e.g. see here:
> [sct-fsleyes_not_enough_inputs.mov.zip](https://github.com/neuropoly/spinalcordtoolbox/files/4794709/sct-fsleyes_not_enough_inputs.mov.zip)

Yes I had noticed this with the initial code on master. I found it weird, but didn't touch it, I assumed maybe I was doing something wrong in my sequence of testing. I will add the error message.

jcohenadad commented:
> > @Drulex currently testing your PR. Overall, the new features look exciting!!! esp. the terminal output embedded in the FSLeyes, i love it!
> > i'm running into a minor issue and i was just wondering if this is expected behaviour: when pressing stop during a process, a window with no message appear, see here:
> > [Untitled.mov.zip](https://github.com/neuropoly/spinalcordtoolbox/files/4794694/Untitled.mov.zip)
> 
> The window shows the `stderr` output if an error occurred. When I try this I get a traceback for a `KeyboardInterrupt` error (see attached)
> [KeyboardInterrupt.mp4.zip](https://github.com/neuropoly/spinalcordtoolbox/files/4794729/KeyboardInterrupt.mp4.zip)
> 
> Maybe its OS related or maybe in that specific sequence you pressed stop too fast?

working now 😊
i tried to reproduce my previous bug 2x but cannot. So let's consider it a non-issue

jcohenadad commented:
@Drulex in terms of space organization, i find there is a lot of "lost space" on the bottom right, how about moving the terminal output to the right, and the function description to the bottom left, like this:

![image](https://user-images.githubusercontent.com/2482071/84943205-08693000-b0b2-11ea-8764-35819ff4823f.png)

Drulex commented:
> @Drulex in terms of space organization, i find there is a lot of "lost space" on the bottom right, how about moving the terminal output to the right, and the function description to the bottom left, like this:
> 
> ![image](https://user-images.githubusercontent.com/2482071/84943205-08693000-b0b2-11ea-8764-35819ff4823f.png)
> 
> it is also more consistent:
> 
>     * column 1: description
> 
>     * column 2: options/inputs
> 
>     * column 3: output

Sure I will change it. I always appreciate UI suggestions because I am terrible at designing. If you have more keep them coming!

Drulex commented:
I changed the layout + cleaned up some stuff along the way:

![image](https://user-images.githubusercontent.com/7751484/85039544-da3d2c00-b155-11ea-9615-89fbc8be1150.png)

jcohenadad commented:
much better! may i suggest additional touch ups:
- in order to see the function description, we need to vertically extend the SCT panel, which leaves very few place to the image. I suggest moving the citation next to the logo (reduce the font to match the size of the terminal output, and if necessary reduce logo size)
- i would also reduce the font size of the function description to match font size of terminal output & buttons
- i don't think we need to write "terminal output", that way, the window is top-vertically aligned with the buttons from the 2nd column

It would look like this:
![image](https://user-images.githubusercontent.com/2482071/85085978-74799000-b1a7-11ea-9c4e-8a716efcf864.png)

Drulex commented:
> much better! may i suggest additional touch ups:
> 
>     * in order to see the function description, we need to vertically extend the SCT panel, which leaves very few place to the image. I suggest moving the citation next to the logo (reduce the font to match the size of the terminal output, and if necessary reduce logo size)
> 
>     * i would also reduce the font size of the function description to match font size of terminal output & buttons
> 
>     * i don't think we need to write "terminal output", that way, the window is top-vertically aligned with the buttons from the 2nd column
> 
> 
> It would look like this:
> ![image](https://user-images.githubusercontent.com/2482071/85085978-74799000-b1a7-11ea-9c4e-8a716efcf864.png)

done!

jcohenadad commented:
@Drulex i tested 2afa6bd, looks cool! minor additional suggestions:
- after giving it some thoughts, it might actually be better to get rid of the box right to the logo, and move the text (General citation...) at the end of the text that starts with "Function description:".
- for the "Function description:" box, can we have the same font type and size as in the terminal output box?
- The help button is truncated:
  <img width="415" alt="Screen Shot 2020-07-07 at 3 49 44 PM" src="https://user-images.githubusercontent.com/2482071/86835211-c5272f00-c069-11ea-8ec3-60ebf8b63350.png">

Drulex commented:
>     * The help button is truncated:

Widgets will be scaled depending on the size of the window or plugin panel. Are you opening the app in fullscreen? What resolution are you using?

I don't see truncation by default, but I can achieve it by scaling the window or the panel (see video attached). It probably has to do with my resolution (1440p).

[fsleyes_scaling.mp4.zip](https://github.com/neuropoly/spinalcordtoolbox/files/4890805/fsleyes_scaling.mp4.zip)

I will try reducing the height of the function description box. The sct script panel already uses half the height available in fsleyes, it would not make sense to increase the height of the panel.

jcohenadad commented:
> > ```
> > * The help button is truncated:
> > ```
> 
> Widgets will be scaled depending on the size of the window or plugin panel. Are you opening the app in fullscreen? What resolution are you using?
> 
> I don't see truncation by default, but I can achieve it by scaling the window or the panel (see video attached). It probably has to do with my resolution (1440p).
> 
> [fsleyes_scaling.mp4.zip](https://github.com/neuropoly/spinalcordtoolbox/files/4890805/fsleyes_scaling.mp4.zip)
> 
> I will try reducing the minimum height of the function description box. The sct script panel already uses half the height available in fsleyes, it would not make sense to increase the height of the panel.

i'm with 1080p. One solution would be to move the help button on the right of the SCT logo (now that the citation has moved in the function window).

Drulex commented:
- Moved help next to logo.
- Moved general citation inside function description (at end).
- Changed function description font and size to match terminal output.

![image](https://user-images.githubusercontent.com/7751484/86932534-7f20a880-c107-11ea-9f89-2f2a5272fd0f.png)

jcohenadad commented:
Looks much better! Two additional comments:
- Until we provide proper documentation on RTD, could you please redirect to [this page](http://forum.spinalcordmri.org/c/sct) when people click on the help?
- Could you please make the arrow on the left the same size as the arrow on the right?
  
<img width="993" alt="Screen Shot 2020-07-08 at 10 48 01 AM" src="https://user-images.githubusercontent.com/2482071/86933851-04f12380-c109-11ea-9af9-a9a17ab6d4fd.png">

jcohenadad commented:
can we merge?

Drulex commented:
> can we merge?

yes
